examples/cpp/main.cpp

```diff
@@ -0,0 +1,316 @@
+     * it in a string (think of the string as an array of bytes). We
+     * run it on a separate thread so that we can wait for the user
+     * to press the Enter key on the main thread to stop it.
+     */
+    std::atomic_bool stopAudio(false);
+    std::string audioBuffer;
+    auto recordAudio = [&stopAudio, &audioBuffer, &config]()
+    {
+        // Start recording
+        Recorder rec(config.recordCmd());
+        rec.start();
+
+        // Collect data
+        while(!stopAudio)
+            audioBuffer += rec.readAudio();
+
+        // Stop recording
+        rec.stop();
+    };
+
+    // Start the thread
+    Timer recordTimer;
+    std::thread audioThread(recordAudio);
+
+    // On the main thread, wait for a keypress to end recording
+    waitForEnter();
+    stopAudio = true;
+    audioThread.join();
+    double recordDuration = recordTimer.elapsed();
+
+    // Send the data to Cubic for the transcription
+    cobaltspeech::cubic::RecognitionConfig rConf = config.recognitionConfig();
+    Timer cubicTimer;
+    cobaltspeech::cubic::RecognitionResponse response =
+        client->recognize(rConf, audioBuffer.c_str(), audioBuffer.length());
+    double cubicDuration = cubicTimer.elapsed();
+
+    // Print the result
+    std::string result = extractTranscript(response);
+    if(result.empty())
+    {
+        std::cout << "Unrecognized result: " << response.Utf8DebugString();
+    }
+    else
+    {
+        std::cout << "Transcription: " << result;
+    }
+    std::cout << std::endl << std::endl;
+
+    // Print out statistics
+    double totalDuration = totalTimer.elapsed();
+    std::cout << "  ASR statistics" << std::endl
+              << "  --------------" << std::endl
+              << "  record time: " << recordDuration << "s" << std::endl
+              << "  cubic time: " << cubicDuration << "s" << std::endl
+              << "    real time factor (record/cubic): "
+              << recordDuration/cubicDuration << std::endl
+              << "  total ASR time: " << totalDuration << "s" << std::endl
+              << std::endl;
+}
+
+/*
+ * Run streaming speech recognition, where results may come back while audio
+ * is still being pushed to the server. This is the recommended approach for
+ * Cubic, particularly if the input audio stream is long.
+ */
+void streamingRecognize(CubicClient *client, const DemoConfig &config)
+{
+    std::cout << "Press ENTER to start recording (Ctrl+C to exit)" << std::endl;
+    waitForEnter();
+    std::cout << "Running ASR (press ENTER to stop)..." << std::endl;
+
+    Timer totalTimer;
+
+    // Create the streaming recognizer
+    cobaltspeech::cubic::RecognitionConfig rConf = config.recognitionConfig();
+    CubicRecognizerStream stream = client->streamingRecognize(rConf);
+
+    // Setup a thread to read audio and push it to Cubic.
+    std::atomic_bool stopAudio(false);
+    auto recordAudio = [&stopAudio, &stream, &config]()
```

Comment:
What is this magic?  I looked up cpp Documentation, but I was still a little lost.  Is it just an anonymous function?

Reply:
It is a [C++ lambda function](https://en.cppreference.com/w/cpp/language/lambda), so yes, it is basically an anonymous function. The arguments in the square brackets are referred to as "captures". The lambda function is capturing these variables from the parent scope, making them available to the function we are defining without having to pass them as function args. It's basically a way to tell the compiler where these variables are coming from (the parent scope). Here I am also passing them by reference instead of value.